### PR TITLE
add Iterators.partition for DataFrameRows

### LINKED ISSUE #3298: When partitioned, partition might lose the missingness eltype (in Tables.schema)
**Problem**: If a user partitions a DataFrame with `Iterators.partition(Tables.rows(df), 2)`, the corresponding `Tables.schema` for each partition will not be type-stable 
Eg, if the first partition does not have any missing information, it would not have the missing type in its schema despite the overall vector having missingness allowed. 

**Why it's a problem**: Arrow.write determines the schema from the first record batch, because Tables-compatible sources retain parents' schema information even for partitions. In effect, missing fields could be lost and replaced by empty fields of the concrete type (eg, "" for String).

**Example**
```
# correct when working with Tables object
t = Tables.rowtable((; x1 =["a","b",missing,"c"], x2 = 1:4))
for part in Iterators.partition(Tables.rows(t), 2)
    @info "Parent type: $(part.parent|>Tables.schema)"
    @info "Columns type: $(Tables.columns(part)|>Tables.schema)"
end

  ┌ Info: Parent type: Tables.Schema:
  │  :x1  Union{Missing, String}
  └  :x2  Int64
  ┌ Info: Columns type: Tables.Schema:
  │  :x1  Union{Missing, String}
  └  :x2  Int64
  ┌ Info: Parent type: Tables.Schema:
  │  :x1  Union{Missing, String}
  └  :x2  Int64
  ┌ Info: Columns type: Tables.Schema:
  │  :x1  Union{Missing, String}
  └  :x2  Int64

# incorrect when working with DataFrame
df = Tables.rowtable((; x1 =["a","b",missing,"c"], x2 = 1:4)) |> DataFrame
for part in Iterators.partition(Tables.rows(df), 2)
    @info "Parent type: $(part.parent|>Tables.schema)"
    @info "Columns type: $(Tables.columns(part)|>Tables.schema)"
end

  ┌ Info: Parent type: Tables.Schema:
  │  :x1  Union{Missing, String}
  └  :x2  Int64
  ┌ Info: Columns type: Tables.Schema:
  │  :x1  String
  └  :x2  Int64
  ┌ Info: Parent type: Tables.Schema:
  │  :x1  Union{Missing, String}
  └  :x2  Int64
  ┌ Info: Columns type: Tables.Schema:
  │  :x1  Union{Missing, String}
  └  :x2  Int64
```

It might be expected, I'm not sure what's required by Tables interface. But I'm opening for visibility, the original issue is in https://github.com/apache/arrow-julia/issues/403 , because that's where the downstream error happens.

Versioninfo:
> DataFrames v1.5.0

> Julia Version 1.8.5
> Commit 17cfb8e65ea (2023-01-08 06:45 UTC)
> Platform Info:
> OS: macOS (arm64-apple-darwin21.5.0)
> CPU: 8 × Apple M1 Pro
> WORD_SIZE: 64
> LIBM: libopenlibm
> LLVM: libLLVM-13.0.1 (ORCJIT, apple-m1)
> Threads: 6 on 6 virtual cores

### PR DESCRIPTION
**Pull request by bkamins**
Fixes #3298 

@svilupp - could you please review this PR against your requirements?

@nalimilan - do you think it should be patch or should go to 1.6 release?

### CODE REVIEW

**Comment by bkamins** [comment]
@svilupp - a related decision is https://github.com/JuliaData/DataFrames.jl/issues/3054. Would you need `Tables.rows(df)` and `Tables.colums(df)` to be a valid Tables.jl table, or it is enough for you that they are iterators?

**Comment by svilupp** [comment]
Wow! So fast! Thank you, @bkamins!

This fixes https://github.com/apache/arrow-julia/issues/403

Missigness is now retained through Arrow serialization and the test case succeeds:
```
df = Tables.rowtable((; x1 =["a","b",missing,"c"], x2 = 1:4)) |> DataFrame
Arrow.write(fn, Iterators.partition(Tables.rows(df), 2); compress = nothing);
t=Arrow.Table(fn)
t[:x1]|>eltype  
# Union{Missing, String}
```

Out of curiosity, it's not entirely clear what makes the default fallback lose the type information given that it's views. Is it [this line](https://github.com/JuliaData/DataFrames.jl/blob/530f7e2d618582f46355f8286c01a7315a88b9e2/src/abstractdataframe/iteration.jl#L131) that fixed it?

***
It would be awesome if this could be a patch, because there are other partitioning use cases where people rely on the general fallback for Tables-sources `Iterators.partition(Tables.rows(df),chunksize)` and DataFrames are arguably the most popular source. 
I think there are other packages affected as well (eg, [here](https://github.com/JuliaData/TableOperations.jl/blob/fab3e72e53d1faf976120de360a5e492fe1e746b/src/TableOperations.jl#L378))

The broader context is that most basic users aren't aware that they need to partition their datasets manually to leverage threading in Arrow (which makes Arrow pretty slow in comparison with PyArrow / Polars, see [here](https://github.com/apache/arrow-julia/issues/393)). 

I hope to change that by chunking data by default (PR [here](https://github.com/apache/arrow-julia/pull/400)), which is something that PyArrow does automatically ([reference](https://github.com/apache/arrow-julia/pull/400)).
For the automatic chunking to work, we'd need to have a consistent chunking method for all Tables-sources (hence, the need for the above call to work). I appreciate that we could add some specializations by type but that would required heavier dependencies.

**Comment by bkamins** [comment]
CI error on nightly is unrelated (it is a problem with compilation time and I cannot reproduce it locally)

**Comment by svilupp** [comment]
> @svilupp - a related decision is https://github.com/JuliaData/DataFrames.jl/issues/3054. Would you need Tables.rows(df) and Tables.colums(df) to be a valid Tables.jl table, or it is enough for you that they are iterators?

For the `Arrow.write` use case, I don't think we need more than what we have (take it with a pinch of salt -- it's my first week of getting to know Arrow.jl code base). 
The only Tables methods I see used are: `partitions`, `columns` (and Columns), `schema` (and Schema), `rowcount`, `columnnames`, `getcolumn` and `eachcolumn`.

I think `rowcount` is the equivalent of `nrow` (mentioned in the issue 3054), but it already has a pretty [decent fallback](https://github.com/JuliaData/Tables.jl/blob/e5487a0bc1094e1116f6ef8cd9aebbb08ec9d363/src/fallbacks.jl#L8)

Perhaps @quinnj is better suited to provide his opinion here?

**Comment by bkamins** [comment]
> Out of curiosity, it's not entirely clear what makes the default fallback lose the type information given that it's views.

No, the reason was that previously we relied on Julia inferring the type of elements of the collection by inspecting them (and a narrowest eltype was therefore picked). Now we pass object that carries over schema information.

---

Aside. The pattern:
```
Iterators.partition(Tables.rows(df),chunksize)
```
does not seem to be fully correct. The point is that `Tables.rows` contract does not guarantee to return a valid Tables.jl table. It only guarantees to return an `AbstractRow`-compatible iterator , which is a weaker requirement. However, @quinnj probably should verify this claim. However, this is the reason of your problem. Before this PR an iterator is returned, but its schema is `nothing`.

This is the situation before this PR:
```
julia> isnothing(Tables.schema(first(Iterators.partition(eachrow(df), 2))))
true
```
what the PR changes is that the above statement returns a schema of the underlying data frame (as this is doable, however, I believe that Tables.jl contract does not guarantee this).

File: src/abstractdataframe/iteration.jl [hunk 1]
```diff
@@ -92,6 +92,66 @@ Compat.hasproperty(itr::DataFrameRows, s::AbstractString) = haskey(index(parent(
 # Private fields are never exposed since they can conflict with column names
 Base.propertynames(itr::DataFrameRows, private::Bool=false) = propertynames(parent(itr))
 
+"""
+    Iterators.partition(dfr::DataFrameRows, n::Integer)
+
+Iterate over `dfr` `DataFrameRows` `n` rows at a time, returning each block
+as a `DataFraneRows` over a view of rows of parent of `dfr`.
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl:99]:
```suggestion
as a `DataFrameRows` over a view of rows of parent of `dfr`.
```

File: src/abstractdataframe/iteration.jl [hunk 2]
```diff
@@ -92,6 +92,66 @@ Compat.hasproperty(itr::DataFrameRows, s::AbstractString) = haskey(index(parent(
 # Private fields are never exposed since they can conflict with column names
 Base.propertynames(itr::DataFrameRows, private::Bool=false) = propertynames(parent(itr))
 
+"""
+    Iterators.partition(dfr::DataFrameRows, n::Integer)
+
+Iterate over `dfr` `DataFrameRows` `n` rows at a time, returning each block
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl:98]:
```suggestion
Iterate over `DataFrameRows` `dfr` `n` rows at a time, returning each block
```

File: src/abstractdataframe/iteration.jl [hunk 3]
```diff
@@ -92,6 +92,66 @@ Compat.hasproperty(itr::DataFrameRows, s::AbstractString) = haskey(index(parent(
...
+# use autodetection of eltype
+Base.IteratorEltype(::Type{<:Iterators.PartitionIterator{<:DataFrameRows}}) =
+    Base.EltypeUnknown()
+
+# we do not need to be overly specific here as we rely on autodetection of eltype
+# this method is needed only to override the fallback for `PartitionIterator`
+Base.eltype(::Type{<:Iterators.PartitionIterator{<:DataFrameRows}}) =
+    DataFrameRows
+
+IteratorSize(::Type{<:Iterators.PartitionIterator{<:DataFrameRows}}) =
```
Comment (nalimilan) [inline src/abstractdataframe/iteration.jl:139]:
```suggestion
Base.IteratorSize(::Type{<:Iterators.PartitionIterator{<:DataFrameRows}}) =
```
(But this method probably isn't needed given that you didn't notice this!)

**Review by nalimilan — APPROVED** [review]

File: src/abstractdataframe/iteration.jl (hunk 3 above)
Comment (bkamins) [reply src/abstractdataframe/iteration.jl:139]: It is not needed, but I wanted to have this to be explicit. Thank you for catching this mistake as I did not understand why we had test coverage problem.

**Comment by bkamins** [comment]
Thank you!

**Pull request merged**